### PR TITLE
[BEAM-59] Remove IOChannelUtils from PackageUtil

### PR DESCRIPTION
**Pull request by vikkyrk**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by vikkyrk** [comment]
R: @dhalperi

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java [hunk 1]
```diff
@@ -110,7 +107,8 @@ static PackageAttributes createPackageAttributes(File source,
 
       // Create the DataflowPackage with staging name and location.
       String uniqueName = getUniqueContentName(source, hash);
-      String resourcePath = IOChannelUtils.resolve(stagingPath, uniqueName);
+      String resourcePath = FileSystems.matchSingleFileSpec(stagingPath).resourceId()
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: I think this should be `matchNewResource`. Staging dir may not exist, for example.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/PackageUtilTest.java [hunk 2]
```diff
@@ -202,18 +238,19 @@ public void testPackageNamingWithDirectoriesHavingSameContentsButDifferentNames(
 
   @Test
   public void testPackageUploadWithLargeClasspathLogsWarning() throws Exception {
-    File tmpFile = makeFileWithContents("file.txt", "This is a test!");
-    // all files will be present and cached so no upload needed.
-    when(mockGcsUtil.fileSize(any(GcsPath.class))).thenReturn(tmpFile.length());
+    File tmpFile = makeFileWithContents("userPackageName=file.txt", "This is a test!");
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/PackageUtilTest.java]: why the change?

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/PackageUtilTest.java [hunk 3]
```diff
@@ -222,20 +259,23 @@ public void testPackageUploadWithFileSucceeds() throws Exception {
...
+
     when(mockGcsUtil.create(any(GcsPath.class), anyString())).thenReturn(pipe.sink());
 
     List<DataflowPackage> targets = PackageUtil.stageClasspathElements(
-        ImmutableList.of(tmpFile.getAbsolutePath()), STAGING_PATH, mockGcsUtil);
+        ImmutableList.of(tmpFile.getAbsolutePath()), STAGING_PATH);
     DataflowPackage target = Iterables.getOnlyElement(targets);
 
-    verify(mockGcsUtil).fileSize(any(GcsPath.class));
+    verify(mockGcsUtil, times(2)).getObjects(anyListOf(GcsPath.class));
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/PackageUtilTest.java]: suspect these are because of `matchSingleFileSpec` vs `matchNewResource`.

**Comment by vikkyrk** [comment]
Retest this please

**Comment by dhalperi** [comment]
Looks to me like the loss of buffer size control has broken things. I believe the intent was to be able to use the CreateOptions to pass this type of info in. Take a look?

**Comment by dhalperi** [comment]
```
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at org.apache.beam.runners.dataflow.repackaged.com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:111)
	at org.apache.beam.runners.dataflow.repackaged.com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:58)
	at org.apache.beam.runners.dataflow.repackaged.com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:75)
	... 3 more
Caused by: java.lang.IllegalArgumentException: Self-suppression not permitted
	at java.lang.Throwable.addSuppressed(Throwable.java:1043)
	at org.apache.beam.runners.dataflow.util.PackageUtil.stageOnePackage(PackageUtil.java:220)
	... 9 more
Caused by: java.lang.OutOfMemoryError: Java heap space
	at com.google.api.client.googleapis.media.MediaHttpUploader.setContentAndHeadersOnCurrentRequest(MediaHttpUploader.java:603)
	at com.google.api.client.googleapis.media.MediaHttpUploader.resumableUpload(MediaHttpUploader.java:409)
	at com.google.api.client.googleapis.media.MediaHttpUploader.upload(MediaHttpUploader.java:336)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:427)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469)
	at com.google.cloud.hadoop.util.AbstractGoogleAsyncWriteChannel$UploadOperation.call(AbstractGoogleAsyncWriteChannel.java:357)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	... 3 more
```

**Comment by vikkyrk** [comment]
Yes taking a look.

**Comment by vikkyrk** [comment]
The test is passing when run locally using `mvn test -pl examples/java -Dtest=WordCountIT -DskipITs=false -DbeamTestPipelineOptions='["--tempRoot=...", "--project=...", "--runner=org.apache.beam.runners.dataflow.testing.TestDataflowRunner"]' -P dataflow-runner` , but fails on jenkins. Most likely issue is that my PR creates a new gcs client for every file which wasn't the case previously. Passing an instance of `GcsUtil` via `GcsCreateOptions` might be a better option instead of passing individual options like `gcsUploadBufferSize` and creating the client in the `GcsFS`

**Comment by dhalperi** [comment]
Do you think it's possible to plumb the options from the create channel request to here? https://github.com/apache/beam/blob/master/sdks/java/extensions/gcp-core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java#L429

**Comment by vikkyrk** [comment]
That should work!. Fingers crossed on tests.

**Comment by vikkyrk** [comment]
Retest this please (unrelated python failures)

File: sdks/java/extensions/gcp-core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java [hunk 4]
```diff
@@ -436,8 +436,16 @@ public SeekableByteChannel open(GcsPath path)
...
-      String type) throws IOException {
+  public WritableByteChannel create(GcsPath path, String type) throws IOException {
+    return create(path, type, uploadBufferSizeBytes);
+  }
+
+  /**
+   * Same as {@link GcsUtil#create(GcsPath, String)} but allows overriding
+   * {code uploadBufferSizeBytes}.
+   */
+  public WritableByteChannel create(GcsPath path, String type, Integer uploadBufferSizeBytes)
```
Comment (dhalperi) [inline sdks/java/extensions/gcp-core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java]: Can't this take CreateOptions directly? That way the next option does not need to be plumbed.

Comment (vikkyrk) [inline sdks/java/extensions/gcp-core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java]: That would add a dependency of `sdk.io.fs` to `sdk.util.GcsUtil`, and there are direct usages of `GcsUtil`, so I wasn't sure.

Comment (vikkyrk) [inline sdks/java/extensions/gcp-core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java]: One option is to move `GcsCreateOptions` to `sdks/java/extensions/gcp-core` (where `GcsUtil` resides), and that way there won't be circular dependencies between `sdks/java/gcp` and `sdks/java/extensions/gcp-core`. But I am still hesitant to have `GcsUtil` depend on any `sdk.io.fs` just for create options, maybe worth taking a look later into unifying options for all the operations?

**Comment by vikkyrk** [comment]
Retest this please
